### PR TITLE
Class imbalance warnings

### LINKED ISSUE #271: Add class imbalance warnings

### PR DESCRIPTION
**Pull request by gabegma**
Resolve #271 

## Description:
Screen Shot 2022-10-18 at 9 37 46 PM
Screen Shot 2022-10-18 at 9 37 39 PM

## Checklist:

You should check all boxes before the PR is ready. If a box does not apply, check it to acknowledge it.

* [x] **ISSUE NUMBER.** You linked the issue number (Ex: Resolve #XXX).
* [x] **PRE-COMMIT.** You ran pre-commit on all commits, or else, you
  ran `pre-commit run --all-files` at the end.
* [x] **USER CHANGES.** The changes are added to CHANGELOG.md and the documentation, if they impact
  our users.
* [x] **DEV CHANGES.**
    * Update the documentation if this PR changes how to develop/launch on the app.
    * Update the `README` files and our wiki for any big design decisions, if relevant.
    * Add unit tests, docstrings, typing and comments for complex sections.

### CODE REVIEW

**Comment by gabegma** [comment]
I'll fix the test tomorrow.

File: azimuth/plots/dataset_warnings.py [hunk 1]
```diff
@@ -167,6 +165,123 @@ def min_nb_samples_plot(
...
+
+    """
+    order = eval_nb_sample.argsort()
+
+    fig = nb_samples_plot(
+        train_nb_sample, eval_nb_sample, alert_train, alert_eval, cls_names, order
+    )
+    num_classes = len(cls_names)
+    common_args = dict(layer="below", line_width=1, y0=-1)
+    common_args_line = dict(type="line", line_dash="dot", **common_args)
```
Comment (JosephMarinier) [inline azimuth/plots/dataset_warnings.py:245]: So, the lines for the mean end up behind the rectangles? Have you tried if that looks better than adding them over the rectangles?

File: azimuth/plots/dataset_warnings.py [hunk 2]
```diff
@@ -167,6 +165,123 @@ def min_nb_samples_plot(
...
+        **common_args_rect,
+    )
+    fig.add_shape(
+        x0=train_mean - max_perc_delta * train_mean,
+        x1=train_mean + max_perc_delta * train_mean,
+        y1=num_classes + 1,
+        fillcolor=Colors.DataViz2,
+        line_color=Colors.DataViz1,
+        **common_args_rect,
+    )
```
Comment (JosephMarinier) [inline azimuth/plots/dataset_warnings.py:276]: Detail, but you could avoid a little bit of repetition of the `mean` variables (and therefore the potential for mistakes) with something like:
```suggestion
    fig.add_shape(
        x0=eval_mean * (1 - max_perc_delta),
        x1=eval_mean * (1 + max_perc_delta),
        y1=num_classes + 2,
        fillcolor=Colors.DataViz1,
        line_color=Colors.DataViz1,
        **common_args_rect,
    )
    fig.add_shape(
        x0=train_mean * (1 - max_perc_delta),
        x1=train_mean * (1 + max_perc_delta),
        y1=num_classes + 1,
        fillcolor=Colors.DataViz2,
        line_color=Colors.DataViz1,
        **common_args_rect,
    )
```

File: azimuth/plots/dataset_warnings.py [hunk 3]
```diff
@@ -167,6 +165,123 @@ def min_nb_samples_plot(
...
+    fig = nb_samples_plot(
+        train_nb_sample, eval_nb_sample, alert_train, alert_eval, cls_names, order
+    )
+    num_classes = len(cls_names)
+    common_args = dict(layer="below", line_width=1, y0=-1)
+    common_args_line = dict(type="line", line_dash="dot", **common_args)
+    fig.add_shape(
+        x0=eval_mean,
+        x1=eval_mean,
+        y1=num_classes + 2,
```
Comment (JosephMarinier) [inline azimuth/plots/dataset_warnings.py:249]: You could extract some constants like those and they would be used 3 times each.
```py
eval_y = num_classes + 2
train_y = num_classes + 1
```

File: azimuth/plots/dataset_warnings.py [hunk 4]
```diff
@@ -167,6 +165,123 @@ def min_nb_samples_plot(
...
+        fillcolor=Colors.DataViz1,
+        line_color=Colors.DataViz1,
+        **common_args_rect,
+    )
+    fig.add_shape(
+        x0=train_mean - max_perc_delta * train_mean,
+        x1=train_mean + max_perc_delta * train_mean,
+        y1=num_classes + 1,
+        fillcolor=Colors.DataViz2,
+        line_color=Colors.DataViz1,
```
Comment (JosephMarinier) [inline azimuth/plots/dataset_warnings.py:274]: Do those `line_color`s have any effect? If so, is this a typo?
```suggestion
        line_color=Colors.DataViz2,
```

File: azimuth/modules/dataset_analysis/dataset_warnings.py [hunk 5]
```diff
@@ -80,6 +81,15 @@ def get_general_warnings(
         alert_eval_nb_min: List[bool] = eval_dist < min_num_per_class
         alert_nb_min: List[bool] = alert_train_nb_min + alert_eval_nb_min
 
+        train_mean = np.mean(train_dist)
+        eval_mean = np.mean(eval_dist)
+        perc_imb_train = (train_dist - train_mean) / train_mean
+        perc_imb_eval = (eval_dist - eval_mean) / eval_mean
```
Comment (JosephMarinier) [inline azimuth/modules/dataset_analysis/dataset_warnings.py:87]: Similarly
```suggestion
        perc_imb_train = train_dist / train_mean - 1
        perc_imb_eval = eval_dist / eval_mean - 1
```

File: azimuth/modules/dataset_analysis/dataset_warnings.py [hunk 6]
```diff
@@ -114,6 +124,41 @@ def get_general_warnings(
...
+                        data=[
+                            DatasetDistributionComparisonValue(
+                                value=perc_imb_train[i], alert=alert_imb_train[i]
+                            ),
+                            DatasetDistributionComparisonValue(
+                                value=perc_imb_eval[i], alert=alert_imb_eval[i]
+                            ),
+                        ],
+                    )
+                    for i, _ in enumerate(alert_imb)
```
Comment (JosephMarinier) [inline azimuth/modules/dataset_analysis/dataset_warnings.py:152]: I think the most pythonic way (and supposedly the most efficient way) to index all those lists would be to use `zip`. That being said, it is clearly insignificant here, the other `DatasetWarning`s are not doing that either, and I personally don't particularly like 😝. So up to you, I just thought it was interesting.
```suggestion
                    DatasetDistributionComparison(
                        name=name,
                        alert=alert,
                        data=[
                            DatasetDistributionComparisonValue(value=val_train, alert=alert_train),
                            DatasetDistributionComparisonValue(value=val_eval, alert=alert_eval),
                        ],
                    )
                    for name, alert, val_train, alert_train, val_eval, alert_eval in zip(
                        class_names,
                        alert_imb,
                        perc_imb_train,
                        alert_imb_train,
                        perc_imb_eval,
                        alert_imb_eval,
                    )
```

File: azimuth/modules/dataset_analysis/dataset_warnings.py [hunk 7]
```diff
@@ -80,6 +81,15 @@ def get_general_warnings(
...
         alert_nb_min: List[bool] = alert_train_nb_min + alert_eval_nb_min
 
+        train_mean = np.mean(train_dist)
+        eval_mean = np.mean(eval_dist)
+        perc_imb_train = (train_dist - train_mean) / train_mean
+        perc_imb_eval = (eval_dist - eval_mean) / eval_mean
+        max_delta_imb = self.config.dataset_warnings.max_delta_class_imbalance
+        alert_imb_train: List[bool] = np.abs(perc_imb_train) > max_delta_imb
+        alert_imb_eval: List[bool] = np.abs(perc_imb_eval) > max_delta_imb
+        alert_imb: List[bool] = alert_imb_train + alert_imb_eval
```
Comment (JosephMarinier) [inline azimuth/modules/dataset_analysis/dataset_warnings.py:91]: Numpy seems to support a real OR operator, as documented [here](https://numpy.org/doc/stable/reference/generated/numpy.logical_or.html#:~:text=The%20%7C%20operator%20can%20be%20used%20as%20a%20shorthand%20for%20np.logical_or%20on%20boolean%20ndarrays.):
```suggestion
        alert_imb: List[bool] = alert_imb_train | alert_imb_eval
```

Comment (JosephMarinier) [reply azimuth/modules/dataset_analysis/dataset_warnings.py:91]: By the way, are those 3 `List[bool]`s necessary? Isn't it able to infer the type automatically?

**Review by JosephMarinier — APPROVED** [review]
Very well done! Good reuse of code to avoid duplication. 👍 I just have a couple of tiny comments, but that seems to be good to go!

File: azimuth/plots/dataset_warnings.py (hunk 1 above)
Comment (gabegma) [reply azimuth/plots/dataset_warnings.py:245]: Right, let me try!

File: azimuth/plots/dataset_warnings.py (hunk 4 above)
Comment (gabegma) [reply azimuth/plots/dataset_warnings.py:274]: Good catch, typo! I thought they did a border around the rectangle, but I'm not sure anymore. I'll double-check.

File: azimuth/modules/dataset_analysis/dataset_warnings.py (hunk 5 above)
Comment (gabegma) [reply azimuth/modules/dataset_analysis/dataset_warnings.py:87]: However, it makes it a bit less intuitive that this is a relative difference, no?

File: azimuth/modules/dataset_analysis/dataset_warnings.py (hunk 6 above)
Comment (gabegma) [reply azimuth/modules/dataset_analysis/dataset_warnings.py:152]: Interesting to know! I don't like it either. I will keep it as is for now, to keep consistency with the others.

File: azimuth/modules/dataset_analysis/dataset_warnings.py (hunk 7 above)
Comment (gabegma) [reply azimuth/modules/dataset_analysis/dataset_warnings.py:91]: No! They used to be, but `mypy` no longer complains. Good catch!!

File: azimuth/plots/dataset_warnings.py (hunk 4 above)
Comment (gabegma) [reply azimuth/plots/dataset_warnings.py:274]: They didn't do anything :)

File: azimuth/plots/dataset_warnings.py (hunk 1 above)
Comment (gabegma) [reply azimuth/plots/dataset_warnings.py:245]: It does look better, great suggestion

File: azimuth/plots/dataset_warnings.py (hunk 4 above)
Comment (gabegma) [reply azimuth/plots/dataset_warnings.py:274]: Oh no, they did!

**Pull request merged**